kbfsedits/tlf_history.go

```diff
@@ -0,0 +1,354 @@
+
+	// Unmarshal and sort the new messages.
+	for _, msg := range messages {
+		var revList []NotificationMessage
+		err := json.Unmarshal([]byte(msg), &revList)
+		if err != nil {
+			// The messages might be from a new version we don't
+			// understand, so swallow the error.
+			continue
+		}
+
+		for j := len(revList) - 1; j >= 0; j-- {
+			revMsg := revList[j]
+			if revMsg.Version != NotificationV2 {
+				// Ignore messages that are too new for us to understand.
+				continue
+			}
+			revMsg.numWithinRevision = j
+			newEdits = append(newEdits, revMsg)
+		}
+	}
+
+	th.lock.Lock()
+	defer th.lock.Unlock()
+	wn, existed := th.byWriter[writerName]
+	if !existed {
+		wn = &writerNotifications{writerName, nil}
+	}
+	oldLen := len(wn.notifications)
+	newEdits = append(newEdits, wn.notifications...)
+	sort.Sort(newEdits)
+	wn.notifications = newEdits.uniquify()
+	if len(wn.notifications) == oldLen {
+		// No new messages.
+		return nil
+	}
+	if !existed {
+		wn.writerName = writerName
+		th.byWriter[writerName] = wn
+	}
+	// Invalidate the cached results.
+	th.computed = false
+	return nil
+}
+
+type fileEvent struct {
+	delete  bool
+	newName string
+}
+
+type recomputer struct {
+	byWriter      map[string]*writerNotifications
+	modifiedFiles map[string]map[string]bool // writer -> file -> bool
+	fileEvents    map[string]fileEvent       // currentName -> ultimate fate
+	numProcessed  map[string]int             // writer name -> num
+}
+
+func newRecomputer() *recomputer {
+	return &recomputer{
+		byWriter:      make(map[string]*writerNotifications),
+		modifiedFiles: make(map[string]map[string]bool),
+		fileEvents:    make(map[string]fileEvent),
+		numProcessed:  make(map[string]int),
+	}
+}
+
+// processNotification adds the notification to the recomputer's
+// history if it is a create/modify for a file that hasn't yet been
+// deleted.  If the file is renamed in a future revision, the added
+// notification has the new name of the file.  processNotification
+// should be called with notifications in reverse order of their
+// revision number.
+//
+// It returns true if it has added enough notifications for the given
+// writer, and the caller should not send any more for that writer.
+func (r *recomputer) processNotification(
+	writer string, notification NotificationMessage) (doTrim bool) {
+	filename := notification.Filename
+	r.numProcessed[writer]++
+
+	// If the file is renamed in a future revision, rename it in the
+	// notification.
+	eventFilename := filename
+	event, hasEvent := r.fileEvents[filename]
+	if hasEvent && event.newName != "" {
+		notification.Filename = event.newName
+		filename = event.newName
+	}
+
+	// Keep only the creates and modifies for non-deleted files,
+	// but remember the renames and deletes.
+	switch notification.Type {
+	case NotificationCreate, NotificationModify:
+		// Disregard any file that's already been deleted.
+		if hasEvent && event.delete {
+			return false
```

Comment:
If a file is deleted then recreated, does it come here or hit a different path?

Reply:
> If a file is deleted then recreated, does it come here or hit a different path?

Remember this is processing events backwards.  So we see the re-create first, which gets added to the set of notifications, and _then_ we see the delete.  Everything earlier than the delete matching that filename gets ignored when it hits this code.

Reply:
I see; that makes sense. Thanks!